tools/src/main/java/org/apache/kafka/tools/ProducerPerformance.java

```diff
@@ -221,43 +379,86 @@ public void newWindow() {
-            double recsPerSec = 1000.0 * count / (double) elapsed;
-            double mbPerSec = 1000.0 * this.bytes / (double) elapsed / (1024.0 * 1024.0);
-            int[] percs = percentiles(this.latencies, index, 0.5, 0.95, 0.99, 0.999);
-            System.out.printf("%d records sent, %f records/sec (%.2f MB/sec), %.2f ms avg latency, %.2f ms max latency, %d ms 50th, %d ms 95th, %d ms 99th, %d ms 99.9th.\n",
-                              count,
-                              recsPerSec,
-                              mbPerSec,
-                              totalLatency / (double) count,
-                              (double) maxLatency,
-                              percs[0],
-                              percs[1],
-                              percs[2],
-                              percs[3]);
+    private static final class ProducerPerformanceThread extends Thread {
+        private final Producer producer;
+        private final String topicName;
+        private final long numRecords;
+        private final int recordSize;
+        private final int throughput;
+        private final int valueBound;
+        private final MetricStats metricStats;
+        private final ThreadStats threadStats;
+        private final Random random = new Random();
+
+        ProducerPerformanceThread(Producer producer,
+                                  String topicName,
+                                  long numRecords,
+                                  int recordSize,
+                                  int throughput,
+                                  int valueBound,
+                                  MetricStats metricStats,
+                                  ThreadStats threadStats) {
+            this.producer = producer;
+            this.topicName = topicName;
+            this.numRecords = numRecords;
+            this.recordSize = recordSize;
+            this.throughput = throughput;
+            this.valueBound = valueBound;
+            this.metricStats = metricStats;
+            this.threadStats = threadStats;
         }
 
-        private static int[] percentiles(int[] latencies, int count, double... percentiles) {
-            int size = Math.min(count, latencies.length);
-            Arrays.sort(latencies, 0, size);
-            int[] values = new int[percentiles.length];
-            for (int i = 0; i < percentiles.length; i++) {
-                int index = (int) (percentiles[i] * size);
-                values[i] = latencies[index];
+        @Override
+        public void run() {
+
+            long startMs = System.currentTimeMillis();
+            ThroughputThrottler throttler = new ThroughputThrottler(throughput, startMs);
+            int messagesSinceLastMetricRecord = 0;
+            int metricStatsInterval = (int) numRecords / 20;
+            int skipMessagesForMetricSampling = (int) numRecords / 2;
+            for (int i = 0; i < numRecords; i++) {
+                byte[] payload = newPayload();
```

Comment:
We've moved payload creation into the main loop, which ends up conflating the cost of creating payloads with producer performance. Its debatable whether this is a problem since including it probably gives more realistic values, but as this is a benchmark we probably want to avoid conflating the two. Especially in the case where the value is fixed anyway.

Could we just preallocate these, and in the compression case allocate N of them and round-robin through them, with N large enough to ensure none of the set will be duplicated within the same message set?